grr/server/grr_response_server/gui/api_plugins/metadata.py

```diff
@@ -51,14 +60,546 @@ class ApiGetOpenApiDescriptionHandler(api_call_handler_base.ApiCallHandler):
+      node = node[1:-1]
+      node = node.split(":")[-1]
+      node = f"{{{node}}}"
+
+    return node
+
+  def _SimplifyPath(self, path: str) -> str:
+    """Keep only fixed parts and parameter names from Werkzeug URL patterns.
+
+    The OpenAPI specification requires that parameters are surrounded by { }
+    which are added in _SimplifyPathNode.
+    """
+
+    nodes = path.split("/")
+    simple_nodes = [self._SimplifyPathNode(node) for node in nodes]
+
+    simple_path = '/'.join(simple_nodes)
+
+    return simple_path
+
+  def _GetPathArgsFromPath(self, path: str) -> [str]:
+    """Extract path parameters from a Werkzeug Rule URL."""
+    path_args = []
+
+    nodes = path.split("/")
+    for node in nodes:
+      if len(node) > 0 and node[0] == '<' and node[-1] == '>':
+        simple_node = self._SimplifyPathNode(node)
+        simple_node = simple_node[1:-1]
+        path_args.append(simple_node)
+
+    return path_args
+
+  def _GetTypeName(self, t):
+    if isinstance(t, FieldDescriptor):
+      if t.message_type:
+        return self._GetTypeName(t.message_type)
+      if t.enum_type:
+        return self._GetTypeName(t.enum_type)
+
+      return self._GetTypeName(t.type)
+
+    if isinstance(t, Descriptor):
+      return t.full_name
+
+    if isinstance(t, EnumDescriptor):
+      return t.full_name
+
+    if inspect.isclass(t):
+      return t.__name__
+
+    if isinstance(t, int): # It's a protobuf.Descriptor.type value.
+      return self.proto_primitive_types_names[t]
+
+    return str(t) # Cover "BinaryStream" and None.
+
+  def _SetMetadata(self):
+    oas_version = "3.0.3"
+    self.open_api_obj["openapi"] = oas_version
+
+    # The Info Object "info" field.
+    info_obj = dict()
+    info_obj["title"] = "GRR Rapid Response API"
+    info_obj["description"] = "GRR Rapid Response is an incident response " \
+                              "framework focused on remote live forensics."
+
+    contact_obj = dict()
+    contact_obj["name"] = "GRR GitHub Repository"
+    contact_obj["url"] = "https://github.com/google/grr"
+    info_obj["contact"] = contact_obj
+
+    license_obj = dict()
+    license_obj["name"] = "Apache 2.0"
+    license_obj["url"] = "http://www.apache.org/licenses/LICENSE-2.0"
+    info_obj["license"] = license_obj
+
+    version_dict = version.Version()
+    info_obj["version"] = f"{version_dict['major']}.{version_dict['minor']}." \
+                          f"{version_dict['revision']}." \
+                          f"{version_dict['release']}"
+    self.open_api_obj["info"] = info_obj
+
+    self.open_api_obj["servers"] = []
+    server_obj = dict()
+    server_obj["url"] = "/"
+    server_obj["description"] = "Root path of the GRR API"
+    self.open_api_obj["servers"].append(server_obj)
+
+  def _AddPrimitiveTypesSchemas(self):
+    """Creates OpenAPI schemas for Protobuf primitives and BinaryStream."""
+    int_to_name = self.proto_primitive_types_names
+
+    # protobuf2.TYPE_DOUBLE == 1
```

Comment:
Thank you for pointing this out!

I replaced `self.proto_primitive_types_names` with a dictionary holding all information about primitive types, including `BinaryStream`, not only protobuf primitives. The key of the dictionary is meant to be the "internal identification token" of the type: int constants for protobuf primitive types, and the name of the type for `BinaryStream`.

I needed a dictionary instead of an array because in `_GetTypeName` I need to lookup the name of the protobuf primitives somehow (I need a mapping between the int constants and their names) and a dictionary helps to do this in O(1).

I also still need a separate set `primitive_types_names` because in `_GetSchemaOrReferenceObject` I need to differentiate between primitive and composite types which I do based on their name and the set allows for O(1) search complexity.

Please let me know if you have any suggestions for doing this in a better way.